### PR TITLE
Add Syncing Module

### PR DESCRIPTION
**Pull request by hayesgm**
This patch adds the ability to start syncing from remote nodes. We basically start one "sync" process that's responsible for trying to interpret messages from remote peers to form a coherent chain. We spawn any number of (okay, for now, one) TCP connections to remote nodes so that we can ask for headers and bodies for blocks. As the sync module gets BlockHeaders (and BlockBodies), it checks if it has a sufficient number of confirmations from peers (again, currently one) and if so, it tries to process that block and add it to our local chain. If that works, it asks for more headers / bodies to continue the process. Other than the fact our canonical chain ("Blocktree") isn't stored locally, this should basically outline the entire sync process (short of Fast / Warp).

### CODE REVIEW

**Comment by hayesgm** [comment]
@masonforest This should be ready for review.

File: lib/ex_wire/adapter/tcp.ex [hunk 1]
```diff
@@ -13,19 +19,22 @@ defmodule ExWire.Adapter.TCP do
   @doc """
   Starts an outbound peer to peer connection.
   """
-  def start_link(:outbound, host, port, remote_id) do
-    GenServer.start_link(__MODULE__, %{is_outbound: true, host: host, port: port, remote_id: remote_id, active: false})
+  def start_link(:outbound, host, port, remote_id, subscribers \\ []) do
+    remote_id_hex = remote_id |> Binary.drop(1) |> ExthCrypto.Math.bin_to_hex
+    ident = Binary.take(remote_id_hex, 6) <> "..." <> Binary.take(remote_id_hex, -6)
```
Comment (masonforest) [inline lib/ex_wire/adapter/tcp.ex]: Would it make sense to add a  `Peer` struct and move this function in as the `to_string` function?

File: lib/ex_wire/adapter/tcp.ex [hunk 2]
```diff
@@ -49,14 +58,23 @@ defmodule ExWire.Adapter.TCP do
...
+  `{:packet, packet, %{remote_id: remote_id, ident: ident}}`.
   """
   def handle_call({:subscribe, {_module, _function, _args}=mfa}, _from, state) do
     updated_state = Map.update(state, :subscribers, [mfa], fn subscribers -> [mfa | subscribers] end)
 
     {:reply, :ok, updated_state}
   end
 
+  def handle_call({:subscribe, {:server, server}=server}, _from, state) do
+    updated_state = Map.update(state, :subscribers, [server], fn subscribers -> [server | subscribers] end)
```
Comment (masonforest) [inline lib/ex_wire/adapter/tcp.ex:69]: other options if you like ¯\_(ツ)_/¯

`%{state| subscribers: [server | state.subscribers]}` or `update_in(state, [:subscribers], &([server| &1]))`

File: lib/ex_wire/config.ex [hunk 3]
```diff
@@ -4,12 +4,22 @@ defmodule ExWire.Config do
   """
 
   @port Application.get_env(:ex_wire, :port, 30304)
-  @private_key ExthCrypto.ECIES.ECDH.new_ecdh_keypair() |> Tuple.to_list() |> List.last # Application.get_env(:ex_wire, :private_key)
+  @private_key ( case Application.get_env(:ex_wire, :private_key) do
+    key when is_binary(key) -> key
+    :random -> ExthCrypto.ECIES.ECDH.new_ecdh_keypair() |> Tuple.to_list() |> List.last
```
Comment (masonforest) [inline lib/ex_wire/config.ex:9]: Makes me want `ExthCrypto.randomKeyPair()` :)

File: lib/ex_wire/peer_sup.ex [hunk 4]
```diff
@@ -0,0 +1,52 @@
+defmodule ExWire.PeerSup do
```
Comment (masonforest) [inline lib/ex_wire/peer_sup.ex]: Nitpick but wanna call this `PeerSupervisor`?

**Review by masonforest — APPROVED** [review]
🎉 🎉 🎉 Nice work!

Now we just need to get the correct `stateRoot` 🤔 

Left some comments if you want. Ready to merge at your discretion!

File: lib/ex_wire/config.ex (hunk 3 above)
Comment (hayesgm) [reply lib/ex_wire/config.ex:9]: Yeah, I'll add this once we start to handle key management realistically.

**Comment by hayesgm** [comment]
Cool, that `%ExWire.Struct.Peer{}` really helped clean up the `TCP` module. Going to merge this now.

**Pull request merged**